### PR TITLE
bpo-26552: Fixed case where failing `asyncio.ensure_future` did not close the coroutine

### PR DESCRIPTION
**Pull request by kumaraditya303**
https://bugs.python.org/issue26552

### CODE REVIEW

File: Lib/asyncio/base_events.py [hunk 1]
```diff
@@ -424,7 +424,11 @@ def create_task(self, coro, *, name=None):
 
         Return a task object.
         """
-        self._check_closed()
+        try:
+            self._check_closed()
+        except RuntimeError:
+            coro.close()
+            raise
```
Comment (gvanrossum) [inline Lib/asyncio/base_events.py]: This solution seems too general for the specific issue in the bug. I would have expected that you'd modify `_ensure_future()` to keep track of whether it's called `_wrap_awaitable()` and if so, close the created task if the following code fails. What you do here might close a Future that was not created by `_wrap_awaitable()` and whose owner should be responsible for closing it.

**Comment by carlosdamazio** [comment]
In `Lib/asyncio/tasks.py`, there's an `_enclose_future()`, you might be able to find out whether the function has called `wrap_awaitable` or not. I couldn't get the edge case when there's a task with a loop closed, of which was stated on the bug report. You might be able to flag out the specific edge case and then call `create_task` with that flag and corner the edge.

**Comment by kumaraditya303** [comment]
(Rebased to main)

File: Lib/asyncio/base_events.py (hunk 1 above)
Comment (kumaraditya303) [reply Lib/asyncio/base_events.py]: Fixed

**Comment by arhadthedev** [comment]
\*with hope\* Does it fix [`test_asyncio` random failures on a Windows runner](https://github.com/python/cpython/pull/30801) so it can be enabled back?

**Comment by kumaraditya303** [comment]
> *with hope* Does it fix [`test_asyncio` random failures on a Windows runner](https://github.com/python/cpython/pull/30801) so it can be enabled back?

I fixed that issue separately.

**Review by gvanrossum — APPROVED** [review]
Hi Kumar, I had procrastinated on reviewing this because there's some complicated stuff going on. I finally found the time and noticed that the bpo issue was written somewhat misleadingly, so I've added a clarification there.

I now finally figured out what the real issue is and how to deal with it, and your code is correct:

- If the argument is a `Future`, nothing more needs to be done
- If it is a coroutine (i.e. `async def`), we need to close it if `create_task()` raises `RuntimeError` (which it only does for one reason, i.e. if the event loop is closed)
- If it is an awaitable (i.e. something with an `__await__()` method), it may not have a `close()` method; the wrapper created by `_wrap_awaitable()` doesn't need to be closed

So I'm going to land this now. Thanks for your patience!

**Comment by kumaraditya303** [comment]
@gvanrossum I think this should be backported to 3.10 and 3.9.

**Comment by gvanrossum** [comment]
Yeah, makes sense. Adding the labels now.

**Comment by kumaraditya303** [comment]
I am leaving 3.9 as it is too diverged from main.

**Pull request merged**